### PR TITLE
Add when statements to type checker and resolver

### PR DESCRIPTION
**Pull request by Nick-Mazuk**

### CODE REVIEW

File: rust/parser/src/binary_operator_or_if.rs [hunk 1]
```diff
@@ -10,8 +10,8 @@ pub fn binary_operator_or_if<'a>(
     context: ExpressionContext,
 ) -> impl FnMut(ParserInput<'a>) -> IResult<'a, Expression<'a>> {
     alt((
+        map(when_statement(context), Expression::When),
```
Comment (aaronstanek) [inline rust/parser/src/binary_operator_or_if.rs]: It is not obvious to me why you made this change.

`when` statements, `if` statements, and binary operator expressions should have mutually exclusive syntax, so the order in which we check for their presence should not matter.

File: rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs [hunk 2]
```diff
@@ -1927,4 +2032,117 @@ mod test {
...
+        let expression = parse_test_expression(indoc! {"
+            when #rgb(0, 0, 0) is
+                #rgb(red, green, blue) do red == \"0\"
+        "});
+        let result = translate_parsed_expression_to_generic_expression(&mut schema, expression);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn different_when_cases_can_have_different_content_lengths() {
```
Comment (aaronstanek) [inline rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs]: From the test name, I am not sure what behavior is being asserted. What should be different when cases have different content lengths?

File: rust/typed_ast/src/nodes.rs [hunk 3]
```diff
@@ -109,11 +109,17 @@ pub struct TypedUnaryOperatorExpression<T> {
...
+    Name(String),
+    DefaultCase,
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct TypedWhenCase<T> {
     pub expression_type: T,
     /// case_name == None indicates the default case
-    pub case_name: Option<TypedIdentifierExpression<T>>,
+    pub case_name: TypedWhenCaseName,
```
Comment (aaronstanek) [inline rust/typed_ast/src/nodes.rs:122]: This is a much better way to do it. Thank you for making the change.

**Review by aaronstanek — APPROVED** [review]
Approved.

This PR contains a test that I think could benefit from a more descriptive name.

File: rust/parser/src/binary_operator_or_if.rs (hunk 1 above)
Comment (Nick-Mazuk) [reply rust/parser/src/binary_operator_or_if.rs]: I agree that it seems like it shouldn't matter, but it did matter. I did not take much time to figure out why.

Comment (Nick-Mazuk) [reply rust/parser/src/binary_operator_or_if.rs]: Ah you know what it probably is: `when` and `do` aren't considered keywords, so the parser probably believes they are identifiers and start malformed binary operators. By switching the order I'm changing the precedence.

https://github.com/cambiata-team/buri/blob/e9a9fa07dc5ee62ea7403502308270a13da2f3d0/rust/parser/src/is_keyword.rs#L1-L3

I can debug this later today.

File: rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs (hunk 2 above)
Comment (Nick-Mazuk) [reply rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs]: Each when case is a tag. Each of these tags may have content of different lengths. This test ensures that behavior is allowed.

File: rust/parser/src/binary_operator_or_if.rs (hunk 1 above)
Comment (Nick-Mazuk) [reply rust/parser/src/binary_operator_or_if.rs]: Yep, that was the root cause. Fixed it.

File: rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs (hunk 2 above)
Comment (Nick-Mazuk) [reply rust/type_checker/translator/src/parsed_expression_to_generic_expression.rs]: Clarified the test name.

**Pull request closed without merging**